packages/server/src/arpa_reporter/lib/audit-report.js

```diff
@@ -125,22 +132,29 @@ async function getProjectSummaryGroupedByProjectRow(data) {
             `${reportingPeriodEndDate} Total Expenditures for Awards Greater or Equal to $50k`,
             `${reportingPeriodEndDate} Total Aggregate Obligations`,
             `${reportingPeriodEndDate} Total Obligations for Awards Greater or Equal to $50k`,
-        ].forEach((columnName) => { row[columnName] = 0; });
+        ].forEach((columnName) => { projectSummaryV2Row[columnName] = 0; });
     });
 
-    row['Capital Expenditure Amount'] = 0;
+    projectSummaryV2Row['Capital Expenditure Amount'] = 0;
 
     // set values in each column
     records.forEach(async (r) => {
+        // for project summaries v2 report
         const reportingPeriodEndDate = reportingPeriods.filter((reportingPeriod) => r.upload.reporting_period_id === reportingPeriod.id)[0].end_date;
-        row[`${reportingPeriodEndDate} Total Aggregate Expenditures`] += (r.content.Total_Expenditures__c || 0);
-        row[`${reportingPeriodEndDate} Total Aggregate Obligations`] += (r.content.Total_Obligations__c || 0);
-        row[`${reportingPeriodEndDate} Total Obligations for Awards Greater or Equal to $50k`] += (r.content.Award_Amount__c || 0);
-        row[`${reportingPeriodEndDate} Total Expenditures for Awards Greater or Equal to $50k`] += (r.content.Expenditure_Amount__c || 0);
-        row['Capital Expenditure Amount'] += (r.content.Total_Cost_Capital_Expenditure__c || 0);
+        projectSummaryV2Row[`${reportingPeriodEndDate} Total Aggregate Expenditures`] += (r.content.Total_Expenditures__c || 0);
+        projectSummaryV2Row[`${reportingPeriodEndDate} Total Aggregate Obligations`] += (r.content.Total_Obligations__c || 0);
+        projectSummaryV2Row[`${reportingPeriodEndDate} Total Obligations for Awards Greater or Equal to $50k`] += (r.content.Award_Amount__c || 0);
+        projectSummaryV2Row[`${reportingPeriodEndDate} Total Expenditures for Awards Greater or Equal to $50k`] += (r.content.Expenditure_Amount__c || 0);
+        projectSummaryV2Row['Capital Expenditure Amount'] += (r.content.Total_Cost_Capital_Expenditure__c || 0);
+
+        // for kpi report
+        const currentPeriodExpenditure = r.content.Current_Period_Expenditures__c || 0;
+        kpiRow['Number of Subawards'] += (r.type === 'awards50k');
+        kpiRow['Number of Expenditures'] += (currentPeriodExpenditure > 0);
+        kpiRow['Evidence based total spend'] += (r.content.Spending_Allocated_Toward_Evidence_Based_Interventions || 0);
     });
 
-    return row;
+    return [projectSummaryV2Row, kpiRow];
 }
```

Comment:
Nit: I think this functionality makes sense, but I would suggest considering splitting it into two functions?

I think your solution here is technically more efficient, since you only have to loop over the data once.  However, it's a little confusing to read; it's one function generating data for two sheets, you have to "unzip" the data afterward, etc.  And I am very doubtful that this loop is actually a bottleneck for performance.

Reply:
I thought the bottleneck is making the db queries, within asyncBatch? I wanted to avoid making additional calls hence this approach. The two sheets are using the same data too, just surfacing different rows so I felt it made sense to group together.

If you or others feel strongly though, I can separate haha. By two functions, do we mean have one `getProjectSummaryGroupedByProjectRow` and one `getKPIDataGroupedByProjectRow` method? Or do we mean keep the logic but instead of inlining, refactor into a helper function?

Reply:
I don't feel very strongly about this one!  It's definitely okay as it is.

---

I guess I was thinking about splitting apart `getProjectSummaryGroupedByProjectRow`, but keeping a single `createProjectSummariesGroupedByProject`.  I don't actually think `getProjectSummaryGroupedByProjectRow` needs to be async at all, so at least that asyncBatch call should be redundant?

In that case, you might end up with something like this:

```diff
 async function createReportsGroupedByProject(periodId) {
     const records = await recordsForProject(periodId);
     const recordsByProject = getRecordsByProject(records);
     const reportingPeriods = await getAllReportingPeriods();
 
-    const inputs = [];
-
-    Object.entries(recordsByProject).forEach(([projectId, r]) => {
-        inputs.push({ projectId, records: r, reportingPeriods });
-    });
-
-    const reportDataGroupedByProjectData = await asyncBatch(inputs, getReportDataGroupedByProjectRow, 2);
-    const projectSummaryGroupedByProject = reportDataGroupedByProjectData.map((row) => row[0]);
-    const KPIDataGroupedByProject = reportDataGroupedByProjectData.map((row) => row[1]);
+    const projectSummaryGroupedByProject = getProjectSummaryGroupedByProject(reportingPeriods, recordsByProject);
+    const kpiDataGroupedByProject = getKpiDataGroupedByProject(reportingPeriods, recordsByProject);
 
     return [projectSummaryGroupedByProject, KPIDataGroupedByProject];
 }
```

With all of that said, I think it makes a ton of sense to wait until we have some more concrete performance data so we better understand where our real bottlenecks are before doing any meaningful refactoring.